groups/etcd_group_service.go

```diff
@@ -0,0 +1,215 @@
+		return nil, err
+	}
+	if etcdRes.Count == 0 {
+		return nil, constants.ErrGroupNotFound
+	}
+
+	payload := &EtcdGroupPayload{}
+	if err = json.Unmarshal(etcdRes.Kvs[0].Value, payload); err != nil {
+		return nil, err
+	}
+	return payload, nil
+}
+
+func putGroupPayload(ctx context.Context, groupName string, payload *EtcdGroupPayload) error {
+	jsonPayload, err := json.Marshal(payload)
+	if err != nil {
+		return err
+	}
+
+	if payload.LeaseID != 0 {
+		_, err = clientInstance.Put(ctx, groupKey(groupName), string(jsonPayload), clientv3.WithLease(payload.LeaseID))
+	} else {
+		_, err = clientInstance.Put(ctx, groupKey(groupName), string(jsonPayload))
+	}
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// GroupCreate creates a group struct inside ETCD, without TTL
+func (c *EtcdGroupService) GroupCreate(ctx context.Context, groupName string) error {
+	etcdRes, err := clientInstance.Get(ctx, groupKey(groupName), clientv3.WithCountOnly())
+	if err != nil {
+		return err
+	}
+	if etcdRes.Count != 0 {
+		return constants.ErrGroupAlreadyExists
+	}
+	return putGroupPayload(ctx, groupName, &EtcdGroupPayload{})
+}
+
+// GroupCreateWithTTL creates a group struct inside ETCD, with TTL, using leaseID
+func (c *EtcdGroupService) GroupCreateWithTTL(ctx context.Context, groupName string, ttlTime time.Duration) error {
+	etcdRes, err := clientInstance.Get(ctx, groupKey(groupName), clientv3.WithCountOnly())
+	if err != nil {
+		return err
+	}
+	if etcdRes.Count != 0 {
+		return constants.ErrGroupAlreadyExists
+	}
+
+	lease, err := clientInstance.Grant(ctx, int64(ttlTime.Seconds()))
+	if err != nil {
+		return err
+	}
+	return putGroupPayload(ctx, groupName, &EtcdGroupPayload{LeaseID: lease.ID})
+}
+
+// GroupMembers returns all member's UIDs
+func (c *EtcdGroupService) GroupMembers(ctx context.Context, groupName string) ([]string, error) {
+	payload, err := getGroupPayload(ctx, groupName)
+	if err != nil {
+		return nil, err
+	}
+	return payload.Uids, nil
+}
+
+// GroupContainsMember check whether a UID is contained in current group or not
+func (c *EtcdGroupService) GroupContainsMember(ctx context.Context, groupName, uid string) (bool, error) {
+	payload, err := getGroupPayload(ctx, groupName)
+	if err != nil {
+		return false, err
+	}
+
+	_, contains := elementIndex(payload.Uids, uid)
+
+	return contains, nil
+}
+
+// GroupAddMember adds UID to group
+func (c *EtcdGroupService) GroupAddMember(ctx context.Context, groupName, uid string) error {
+	payload, err := getGroupPayload(ctx, groupName)
+	if err != nil {
+		return err
+	}
+
+	_, contains := elementIndex(payload.Uids, uid)
+	if contains {
+		return constants.ErrMemberAlreadyExists
+	}
+
+	payload.Uids = append(payload.Uids, uid)
+	return putGroupPayload(ctx, groupName, payload)
+}
+
+// GroupRemoveMember removes specified UID from group
+func (c *EtcdGroupService) GroupRemoveMember(ctx context.Context, groupName, uid string) error {
+	payload, err := getGroupPayload(ctx, groupName)
```

Comment:
I guess there's a running condition that can happen here?

GroupAddMember and GroupRemoveMember are not atomic, both of them are doing the following:

read
process
write

If another operation begins while one is in the process phase we will have an undefined behaviour.

Reply:
Understood the problem! Changed the implementation to be everything atomic, so that these types of errors don't occur